storage-transcribe-audio/functions/src/transcribe-audio.ts

```diff
@@ -0,0 +1,212 @@
+  const taggedTranscription = response.results?.map(
+    (result) =>
+      [result?.channelTag, result?.alternatives?.[0].transcript] as const
+  );
+
+  if (!isTaggedStringArray(taggedTranscription)) {
+    return {
+      state: "failure",
+      warnings,
+      type: FailureType.NULL_TRANSCRIPTION,
+    };
+  }
+
+  // I simplify the transcription compared to the one that's usually given
+  // by the cloud call because, for example, we don't give the option
+  // to request many candidate transcriptions from speech to text.
+  //
+  // However, the simplification doesn't happen for the file uploaded by the API
+  // Should I
+  // (a) just not simplify?
+  // (b) simplify after the upload has succeeded?
+  // (c) simplify, then upload (without harnessing the upload of the cloud API)?
+  // (d) simplify, keeping the mildly inconsistent behavior?
+  const transcription = separateByTags(taggedTranscription);
+
+  logs.logResponseTranscription(transcription);
+  return {
+    state: "success",
+    warnings,
+    transcription,
+  };
+}
+
+export async function transcodeToLinear16AndUpload(
+  {
+    localCopyPath,
+    storageOutputPath,
+  }: { localCopyPath: string; storageOutputPath: string },
+  bucket: Bucket
+): Promise<TranscodeAudioResult> {
+  const probeData: ffmpeg.FfprobeData = await probePromise(localCopyPath);
+  const warnings: WarningType[] = [];
+
+  logs.debug("probe data before transcription:", probeData);
+  const { streams } = probeData;
+
+  if (streams.length === 0) {
+    return {
+      state: "failure",
+      type: FailureType.ZERO_STREAMS,
+      warnings,
+    };
+  }
+
+  if (streams.length !== 1) {
+    warnings.push(WarningType.MORE_THAN_ONE_STREAM);
+  }
+
+  if (streams[0].sample_rate == null) {
+    return {
+      state: "failure",
+      type: FailureType.NULL_SAMPLE_RATE,
+      warnings,
+    };
+  }
+
+  if (streams[0].channels == null) {
+    return {
+      state: "failure",
+      type: FailureType.NULL_CHANNELS,
+      warnings,
+    };
+  }
+
+  const localOutputPath = localCopyPath + TRANSCODE_TARGET_FILE_EXTENSION;
+  logs.debug("transcoding locally");
+  try {
+    await transcodeLocally({
+      inputPath: localCopyPath,
+      outputPath: localOutputPath,
+    });
+  } catch (error: unknown) {
+    const { err, stdout, stderr } = error as {
+      err: any;
+      stdout: string;
+      stderr: string;
+    };
+    return {
+      state: "failure",
+      type: FailureType.FFMPEG_FAILURE,
+      warnings,
+      details: {
+        ffmpegError: err,
+        ffmpegStdout: stdout,
+        ffmpegStderr: stderr,
+      },
+    };
+  }
+  logs.debug("finished transcoding locally");
+
+  logs.debug("uploading transcoded file");
+  const uploadResponse = await bucket.upload(localOutputPath, {
+    destination: storageOutputPath,
+    metadata: { metadata: { isTranscodeOutput: true } },
+  });
+  logs.debug("uploaded transcoded file");
+
+  return {
+    state: "success",
```

Comment:
What if the upload fails? This function feels like it should be separated into 1) transcribe, and 2) upload.

Reply:
Done.